docs/RFCS/20171118_role_based_access_control.md

```diff
@@ -0,0 +1,747 @@
+
+A user `marc` can be added to the role `myrole` without role admin permissions using:
+```
+GRANT myrole TO marc
+```
+
+`marc` is now a "regular" member of the role, inheriting all role privileges. However, `marc`
+is not allowed to modify role membership.
+
+To allow a user to modify role membership, we can use `WITH ADMIN OPTION`:
+```
+GRANT myrole TO marc WITH ADMIN OPTION
+```
+
+`marc` is now a member of the role **and** is allowed to modify role membership.
+
+### Dropping a role
+
+Cockroach administrators can delete a role called `myrole` using:
+```
+DROP ROLE myrole
+```
+
+This fails if `myrole` does not exist. We can instead use:
+```
+DROP ROLE IF EXISTS myrole
+```
+
+Dropping a role will fail if it still has privileges on database objects.
+
+Any role relationships involving this role will be removed automatically.
+
+This applies to dropping users as well: dropping user `A` when `A ∈ myrole` will automatically remove
+the membership.
+
+### Listing roles
+
+We can get a simple list of all roles in the cluster by running:
+```
+SHOW ROLES
+
++--------------+
+| Role         |
++--------------+
+| admin        |
+| myrole       |
+| myotherrole  |
++--------------+
+```
+
+### Listing role memberships
+
+A list of members can be found using:
+```
+SHOW GRANTS ON ROLE myrole
+
++--------+----------- +-------+
+| Role   | User/Role  | Admin |
++--------+------------+-------+
+| myrole | marc       | YES   |
+| myrole | other      | NO    |
++--------+------------+-------+
+```
+
+We will also allow the following variants:
+* lookup all roles: `SHOW GRANTS ON ROLE *`
+* show all memberships for a user: `SHOW GRANT ON ROLE * FOR marc`
+* show user membership in a role: `SHOW GRANT ON ROLE myrole FOR marc`
+
+### Administrators
+
+Before this RFC, there is a single administrative user in a cockroach cluster: the user `root`.
+
+We propose to add a default role: `admin` with a single user: `root`.
+`root` will have admin privileges on the `admin` role, allowing addition/removal of other users.
+
+The `admin` role cannot be deleted, and cannot be empty. The special `root` user cannot be dropped or removed
+from the `admin` role.
+
+All operations that previously checked for the `root` user will now check for membership in the `admin` role.
+
+Clusters created before the existence of roles will have a new `admin` role created automatically and privileges
+for it added where `root` currently has privileges. `root` will keep its individual privileges to allow downgrades
+to older versions.
+
+### Non-enterprise users
+
+Manipulation of roles is an enterprise feature.
+
+Without a valid enterprise license, the following are **not** allowed:
+* creating a role (`CREATE ROLE ...`)
+* dropping a role (`DROP ROLE ...`)
+* adding a member of a role (`GRANT myrole TO ...`)
+* removing a member from a role (`REVOKE myrole TO ...`)
+
+The following will function without an enterprise license:
```

Comment:
How does this interact with non-CCL builds? Is it just the code for managing roles that's CCL'ed?

Reply:
Yeah, but there's some question about code location. This is so similar to users for some bits that splitting it into CCL and OSS is really making life purposefully difficult for ourselves.

Reply:
But the answer is yes: create/drop/grant/revoke for roles will be CCL. Granting privileges to a role will still work as will permission checks (obviously). We do need to make sure we get a nice "you need a CCL build and license" message if you run them in the OSS binary.

I think this is a fair way of doing it but I'm open to suggestions.